---
title: "Merck - QC PBMC 3k"
author: "Noor Sohail"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      df_print: paged
      highlights: pygments
      number_sections: true
      self_contained: true
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
# Turn off Warnings and other console output messages from the whole document
```

```{r, cache=FALSE, message=FALSE}
library(tidyverse)
# library(data.table)
library(knitr)
# library(plyr)
# library(dplyr)
library(glue)
library(Seurat)
library(scales)
library(ggh4x)
library(gridExtra)

library(devtools)

ggplot2::theme_set(theme_light(base_size = 11))
opts_chunk[["set"]](
    cache = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE)

axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2, "cm")
)
```

# Project details

``` {r}
set.seed(2023)

name_project <- "Co-expression of genes"
name_pi <- "Merck"

name_file <- gsub(" ", "_", name_project)
name_file <- gsub("-", "_", name_file)
```

<b>Project goals</b>

The Harvard Chan Bioinformatics Core (HBC) will work with Felix Geist, Merck KGaA, Darmstadt, Germany to perform the following services:

Milestone 1: (estimate: 15-20h)

Identification of methods in R that allow the estimation of co-expression of two genes on single cells using adequate statistical methods that take the sparsity of data into consideration (e.g. zero-inflation)
Identification of suitable dataset (or datasets) to use for benchmarking the software
Agreement on the predefined criteria to be used for analyses endpoints
Milestone 2: (estimate: 25-30h)

Perform a comparison of methods in milestone 1 and benchmark those methods

Milestone 3: (estimate: 20-30h)

Preparation of exemplary use cases for the questions raised in single cell seq datasets derived from human tumors with the best performing method

Milestone 4: (estimate: 10-15h)

Reporting of results as Rmarkdown/quarto/Bookdown document(s) including the code.

<b>Dataset background</b>


```{r}
# Name of column we are doing comparisons for
fill <- "sample"

n_samples <- 2
nrows <- 1
ncols <- 2
```

# QC

These QC metrics are the values from the downloaded dataset.


```{r}
# Load dataset
load(bzfile("../data/seurat_integrated.RData.bz2"))
merged <- seurat_integrated
```

```{r}
# celltype annotation
Idents(merged) <- merged$seurat_clusters
merged <- RenameIdents(object = merged, 
                               "0" = "Naive or memory CD4+ T cells",
                               "1" = "CD14+ monocytes",
                               "2" = "Activated T cells",
                               "3" = "CD14+ monocytes",
                               "4" = "Stressed cells / Unknown",
                               "5" = "CD8+ T cells",
                               "6" = "Naive or memory CD4+ T cells",
                               "7" = "B cells",
                               "8" = "NK cells",
                               "9" = "CD8+ T cells",
                               "10" = "FCGR3A+ monocytes",
                               "11" = "B cells",
                               "12" = "NK cells",
                               "13" = "B cells",
                               "14" = "Conventional dendritic cells",
                               "15" = "Megakaryocytes",
			                   "16" = "Plasmacytoid dendritic cells")
merged$celltype <- Idents(merged)
```

# Sources of variability
To enable meaningful visualization of the data, we apply a minimal normalization to our raw data (log-normalization). We then identify the top most variable genes across the log-normalized data, i.e. those with the greatest variability in expression level from one cell to the next. Finally, we calculate principal components (PCs) based on these top `r n_pcs` most variable genes.

## Highly Variable Genes
We isolate the genes that are responsible for the variation in the data. Feature selection is very important as almost all downstream steps make heavy use of these features (ex. calculating PCA, neighbors, and UMAPs).

```{r}
seurat_norm <- merged
```

## PCA {.tabset}

### Principal Components 1 vs 2
For datasets with a larger number of samples or cells, PC1 and PC2 scores for each sample/cell are plotted usually. Since these PCs explain the most variation in the dataset, the expectation is that the samples/cells that are more similar to each other will cluster together with PC1 and PC2.

```{r fig.height=5, fig.width=6}
Idents(merged) <- merged$sample
DimPlot(merged, reduction = "pca") +
    xlab("PC1") +
    ylab("PC2") +
    ggtitle(glue("{name_project}\nPrincipal Component Analysis")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank(),
        axis.title=element_text(size=8, face="bold")) +
    guides(x = axis, y = axis) +
    theme(axis.line = element_line(arrow = arrow(type = "closed"), size = 1),
        axis.title = element_text(hjust = 0))
```

### Top genes per PC
Here we view the top 50 genes that explain the most variance among the first 4 PCs. If there are a lot of ribosomal or mitochondrial genes among these lists, we would expect to see many stressed/dying cells among the dataset.

```{r}
df <- Loadings(seurat_norm[['pca']])
df <- df %>% data.frame()
df$gene <- rownames(df)
rownames(df) <- NULL

genes_pc <- data.frame(df$gene)
for (i in 1:10) {
    pc <- glue("PC_{i}")
    genes <- df %>% arrange(get(pc))
    genes <- genes$gene
    genes_pc[[pc]] <- genes
}
genes_pc[1:50, 2:5]
```

# UMAP

We use the first 40 PCs to derive reduced UMAP (Uniform Manifold Approximation and Projection) components.

We then use the UMAP reduction to explore our dataset and assess how different variables influence cell clustering. Throughout this report, UMAP representations are split by group and/or samples, to enable checking for potential study-specific clustering. At this stage of the normalization process, we are especially concerned with checking the impact of variables that reflect effects that have no biological relevance to the study.

## Sample distribution {.tabset}

### Sample pooled
```{r fig.height=5, fig.width=6}
DimPlot(seurat_norm, reduction = "umap",group.by = "sample") +
    xlab("UMAP1") +
    ylab("UMAP2") +
    ggtitle(glue("{name_project}\nsample")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank(),
        axis.title=element_text(size=8, face="bold")) +
    guides(x = axis, y = axis) +
    theme(axis.line = element_line(arrow = arrow(type = "closed"), size = 1),
        axis.title = element_text(hjust = 0))
```

### Sample split
```{r fig.height=n_rows*5, fig.width=n_cols*5}
sample_names <- sort(unique(seurat_norm$sample))
palette_sample <- hue_pal()(length(sample_names))

# Get UMAP coordinates
df <- seurat_norm@meta.data
df$umap_1 <- seurat_norm[["umap"]][[]][,1]
df$umap_2 <- seurat_norm[["umap"]][[]][,2]

plot_list <- list()
for (i in 1:length(sample_names)) {
    sample_name <- sample_names[i]
    df_sample <- subset(df, sample == sample_name, select = c("sample", "umap_1", "umap_2"))
    color_sample <- palette_sample[i]

    plot_list[[i]] <- ggplot(data = df, aes(x = umap_1, y = umap_2)) +
        geom_point(color = "lightgray", alpha = 0.5, size=1) +
        geom_point(data = df_sample, aes(x = umap_1, y = umap_2), color = color_sample, size=1) +
        xlab("UMAP1") +
        ylab("UMAP2") +
        theme_classic() +
        ggtitle(glue("{sample_name}")) +
        theme(plot.title = element_text(hjust = 0.5)) +
        theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
            axis.text.y=element_blank(), axis.ticks.y=element_blank(),
            axis.title=element_text(size=8, face="bold")) +
        guides(x = axis, y = axis) +
        theme(axis.line = element_line(arrow = arrow(type = "closed"), size = 1),
            axis.title = element_text(hjust = 0))
}

grid.arrange(
    grobs = plot_list,
    ncol = 2,
    widths = rep(1, 2),
    top = glue("{name_project}\nsample")
)
```

## Celltypes
```{r fig.height=5, fig.width=5}
Idents(seurat_norm) <- "celltype"
p <- DimPlot(seurat_norm, reduction = "umap") +
    xlab("UMAP1") +
    ylab("UMAP2") +
    ggtitle(glue("{name_project}\nSeurat Clusters")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank(),
        axis.title=element_text(size=8, face="bold")) +
    guides(x = axis, y = axis) +
    theme(axis.line = element_line(arrow = arrow(type = "closed"), size = 1),
        axis.title = element_text(hjust = 0))
LabelClusters(p, id = "ident",  fontface = "bold", size = 3, bg.colour = "white", bg.r = .2, force = 0)
```

```{r fig.height=5, fig.width=7}
Idents(seurat_norm) <- "celltype"
p <- DimPlot(seurat_norm, reduction = "umap") +
    xlab("UMAP1") +
    ylab("UMAP2") +
    ggtitle(glue("{name_project}\nSeurat Clusters")) +
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(),
        axis.text.y=element_blank(), axis.ticks.y=element_blank(),
        axis.title=element_text(size=8, face="bold")) +
    guides(x = axis, y = axis) +
    theme(axis.line = element_line(arrow = arrow(type = "closed"), size = 1),
        axis.title = element_text(hjust = 0))
p
```


# Run SAVER

We run SAVER imputation on the top 2000 highly variable genes.

```{r}
# devtools::install_github("mohuangx/SAVER")
library(SAVER)

# Identify the indices of the genes of interest
genes <- VariableFeatures(seurat_norm)

# Un-normalized matrix
data <- seurat_norm$RNA@counts
genes.ind <- which(rownames(data) %in% genes)

# Generate predictions for those genes and return entire dataset
saver.genes <- saver(data, pred.genes = genes.ind, estimates.only = TRUE, ncores=8, pred.genes.only = TRUE)
```

```{r}
# Store output from saver into seurat object and save
seurat_norm@imputed <- saver.saver.genes
saveRDS("saver_seurat.rds")
```